### PR TITLE
New update-nanmean implementation and test for jax frontend

### LINKED ISSUE #11739: nanmean

### PR DESCRIPTION
**Pull request by Lexticc**
close #11739

### CODE REVIEW

**Comment by Lexticc** [comment]
this pr is continuation of pr #11972 and pr #12413 reviewed by @Infrared1029 . would you please assign yourself to it to continue what we started in pr #11972 and pr #12413 with @Infrared1029

**Comment by Lexticc** [comment]
I have considered your (@Infrared1029 ) remarks in in pr #11972 and pr # 12413 . I made every possible check to make sure it passes the tests and this is the outcome
please your (@Infrared1029 ) comments are much valued
thank you a lot

**Comment by Lexticc** [comment]
Hello James this pr #12961 is continuation of the pr #11822 which you (@jkeane508) kindly reviewed two weeks ago
thank you

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py [hunk 1]
```diff
@@ -729,3 +729,53 @@ def test_jax_numpy_std(
...
+):
+    input_dtypes, x, axis = dtype_x_axis
+    where, input_dtypes, test_flags = np_helpers.handle_where_and_array_bools(
+        where=where,
+        input_dtype=input_dtypes,
+        test_flags=test_flags,
+    )
+    np_helpers.test_frontend_function(
+        input_dtypes=input_dtypes,
+        all_aliases=["numpy.nanmean"],
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py:769]: This arg here is throwing an error as an unexpected keyword arg

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Hey,

Happy to continue reviewing :)

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py (hunk 1 above)
Comment (Lexticc) [reply ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py]: Hello James ( @jkeane508 ),
I have removed this line === all_aliases=["numpy.nanmean"], === from the code to see if there would be any difference 
would you please check with me the result of this change
thank you so much

**Comment by Lexticc** [comment]
Hello James, @jkeane508
would you please review the changes you requested
Thank you

**Comment by Lexticc** [comment]
hello James (@jkeane508 )
wishing you all the best in this festive season
hope to hear from you soon

**Comment by Lexticc** [comment]
Dear James (@jkeane508 )
today this page showed that there is a conflict that needs to be resolved
before resolving the conflict only 2 tests failed
after resolving the conflict 27 tests failed
please why had this happened although nothing in the code was changed only I rearranged the  conflicted lines

**Comment by jkeane508** [comment]
Hi @Lexticc,

Apologies I was on leave, will be reviewing ASAP!

File: ivy/functional/frontends/jax/numpy/statistical.py [hunk 2]
```diff
@@ -427,3 +427,17 @@ def ptp(a, axis=None, out=None, keepdims=False):
...
+@handle_jax_dtype
+@to_ivy_arrays_and_back
+def nanmean(a, axis=None, dtype=None, out=None, keepdims=False, *, where=None):
+    axis = tuple(axis) if isinstance(axis, list) else axis
+    if dtype is None:
+        dtype = "float32" if ivy.is_int_dtype(a) else a.dtype
+    if ivy.is_array(where):
+        where = ivy.array(where, dtype=ivy.bool)
+        a = ivy.where(where, a, ivy.full_like(a, ivy.nan), out=out)
+    ret = ivy.nanmean(a, axis=axis, dtype=dtype, keepdims=keepdims, out=out)
```
Comment (jkeane508) [inline ivy/functional/frontends/jax/numpy/statistical.py:441]: Hey, I'm pretty sure that nanmean is currently still experimental which means that it isn't going to be available in the newest release. Because of this, I don't think we can call it directly from the frontend. You would need to implement this as a combination of functions in Ivy's functional API instead.

File: ivy/functional/frontends/jax/numpy/statistical.py [hunk 3]
```diff
@@ -427,3 +427,17 @@ def ptp(a, axis=None, out=None, keepdims=False):
     x = ivy.max(a, axis=axis, keepdims=keepdims)
     y = ivy.min(a, axis=axis, keepdims=keepdims)
     return ivy.subtract(x, y)
+
+
+@handle_jax_dtype
+@to_ivy_arrays_and_back
+def nanmean(a, axis=None, dtype=None, out=None, keepdims=False, *, where=None):
+    axis = tuple(axis) if isinstance(axis, list) else axis
```
Comment (jkeane508) [inline ivy/functional/frontends/jax/numpy/statistical.py:435]: as per [docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.nanmean.html), the default dtype is float64. I would use the link here to just check that all behavior is the same :)

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py [hunk 4]
```diff
@@ -831,18 +831,10 @@ def test_jax_numpy_std(
 def test_jax_numpy_corrcoef(
     dtype_and_x,
     rowvar,
-    frontend,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py:834]: How come these were removed from the test?

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py [hunk 5]
```diff
@@ -831,18 +831,10 @@ def test_jax_numpy_std(
...
-    test_flags,
-    fn_tree,
-    on_device,
 ):
     input_dtypes, x = dtype_and_x
     np_helpers.test_frontend_function(
         input_dtypes=input_dtypes,
-        frontend=frontend,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py:842]: Same here?

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Hey

Just had a look through your PR, see below for comments :)

Thanks,
James

File: ivy/functional/frontends/jax/numpy/statistical.py (hunk 2 above)
Comment (Lexticc) [reply ivy/functional/frontends/jax/numpy/statistical.py]: Hello James (@jkeane508)
 
I replaced the direct use of nanmean function with the definition of the mean=sum of elements not equal to nan divided by the count of the elements that are not nan

File: ivy/functional/frontends/jax/numpy/statistical.py (hunk 3 above)
Comment (Lexticc) [reply ivy/functional/frontends/jax/numpy/statistical.py:438]: I used float64 as stated in the docs

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py (hunk 4 above)
Comment (Lexticc) [reply ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py:834]: the last time i resolved the conflict using the web and this apparently caused some line to be removed
today i used pycharm ide and on my fork every thing is perfect

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py (hunk 5 above)
Comment (Lexticc) [reply ivy_tests/test_ivy/test_frontends/test_jax/test_jax_numpy_statistical.py]: the last time i resolved the conflict using the web and this apparently caused some line to be removed
today i used pycharm ide and on my fork every thing is perfect

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Hey, it looks like there a lint errors inside the files in this PR. I would suggest installing a pre-commit hook to prevent these, instructions to install [here](https://lets-unify.ai/docs/ivy/overview/contributing/setting_up.html#pre-commit). 

There seem to be a lot of changes in the test file, ideally, we only want to see changes to the code that you have implemented.

Thanks,
James

**Comment by Lexticc** [comment]
Hey James ( @jkeane508 )
I have corrected the files in this pr and used pre-commit hook as you advised
thank you

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Hi,

Just had a review of the test logs [here](https://github.com/unifyai/ivy/actions/runs/4779766646/jobs/8497055574?pr=12961). Looks like tests are failing in cases where arrays passed are just `[0.]`. Must be an edge case that needs handling.

Good Luck!

**Comment by Lexticc** [comment]
Hello James ( @jkeane508 )
Please check the attached snapshot of my computer it shows that nanmean passed all tests with 100% pass
I could not figure out why === intelligent-tests-pr / run_tests_1 (pull_request) === failed

attached desktop screen shot : 
all pass with casting_final after commit

**Comment by jkeane508** [comment]
> Hello James ( @jkeane508 ) Please check the attached snapshot of my computer it shows that nanmean passed all tests with 100% pass
> 
> thank you so much for your all efforts
> 
> attached desktop screen shot : all pass with casting_final after commit

Hi @Lexticc,

The package we use for testing the hypothesis does random permutations of test parameters. The testing on GitHub has likely reached this edge case before your local setup. I would aim to solve the test failing on GitHub so that we can merge this PR.

It is likely that your local setup hasn't reached the edge case mentioned in my previous review yet :)

Thanks,
James

**Comment by Lexticc** [comment]
Hello James ( @jkeane508 )
you were right in your comment about the edge cases as after double checking I ran through them on my local ide pycharm so I changed the code to handle these edge cases
 which are :
- zero division when all elements are nan in the chosen axis , or when   _**where**_ argument  is false for all elements in the chosen axis
-  in addition to dtype casting when the input and output dtypes are not the same)

 then I made a new commit and the image of my computer is for the results of the test for the latest new commit after changing the code to handle the edge cases

please I found stated at the end line of ======= intelligent-tests-pr / display_test_results (pull_request) ====== :Details

This PR does not introduce any new test failures! Yippee!

dose this mean it can be merged

pass yippe

**Review by jkeane508 — APPROVED** [review]
Yep, this is looking good to me!

Will go ahead and merge :)
